comms/src/connection_manager/establisher.rs

```diff
@@ -0,0 +1,498 @@
+
+        Ok((
+            PeerConnectionEntry {
+                connection,
+                address: address.clone(),
+            },
+            worker_handle,
+        ))
+    }
+
+    /// Establish a new connection for a peer. A connection may be Inbound
+    pub fn establish_inbound_peer_connection(
+        &self,
+        peer: &Peer<CommsPublicKey>,
+        secret_key: CurveSecretKey,
+    ) -> Result<(PeerConnectionEntry, PeerConnectionJoinHandle)>
+    {
+        // Providing port 0 tells the OS to allocate a port for us
+        let mut address = NetAddress::IP((self.config.host, 0).into());
+        let context = self
+            .new_context_builder()
+            .set_id(peer.node_id.clone())
+            .set_direction(Direction::Inbound)
+            .set_address(address.clone())
+            .set_curve_encryption(CurveEncryption::Server { secret_key })
+            .build()?;
+
+        let connection = PeerConnection::new();
+        let worker_handle = connection.start(context)?;
+        connection
+            .wait_connected_or_failure(self.config.establish_timeout)
+            .or_else(|err| {
+                error!(
+                    target: LOG_TARGET,
+                    "Unable to establish inbound connection for NodeId={}: {:?}", peer.node_id, err
+                );
+                Err(ConnectionManagerError::ConnectionError(err))
+            })?;
+
+        let connection = Arc::new(connection);
+        if let Some(addr) = connection.get_connected_address().map(|a| a.into()) {
+            address = addr;
+        }
+
+        Ok((PeerConnectionEntry { connection, address }, worker_handle))
+    }
+
+    fn new_context_builder(&self) -> PeerConnectionContextBuilder {
+        let config = &self.config;
+
+        let mut builder = PeerConnectionContextBuilder::new()
+            .set_context(&config.context)
+            .set_max_msg_size(config.max_message_size)
+            .set_consumer_address(config.consumer_address.clone())
+            .set_max_retry_attempts(config.max_connect_retries);
+
+        if let Some(ref addr) = config.socks_proxy_address {
+            builder = builder.set_socks_proxy(addr.clone());
+        }
+
+        builder
+    }
+}
+
+//---------------------------------- Connection Attempts --------------------------------------------//
+
+use crate::connection::{
+    connection::EstablishedConnection,
+    monitor::{ConnectionMonitor, SocketEventType},
+};
+use tari_storage::keyvalue_store::DataStore;
+
+/// Helper struct which enables multiple attempts at connecting. This also updates the peers connection
+/// attempts statistics
+struct ConnectionAttempts<'c, F, DS> {
+    context: &'c Context,
+    num_attempts: usize,
+    attempt_fn: F,
+    peer_manager: Arc<PeerManager<CommsPublicKey, DS>>,
+}
+
+impl<'c, F, DS> ConnectionAttempts<'c, F, DS>
+where
+    F: Fn(usize, InprocAddress) -> Result<(EstablishedConnection, NetAddress)>,
+    DS: DataStore,
```

Comment:
The datastore generic `DS` could be replaced with `CommsDataStore` from `types.rs`

Reply:
Will do!